---
title: "Sequencing Metrics"
output: pdf_document
---

The following data is taken from the _SequencingMetrics.txt_ file located on ARC in the following location:
_/nobackup/medlste/data/RNAseq/PvR_consolidated/SequencingMetrics.txt_. From the original set of sequencing metrics (183 samples across 92 patients) I removed the following samples:

- Those which are absent from the metadata file (10 samples across 5 patients). Agreed in previous meeting that these should not be included in analysis.
- Those which we previously agreed to filter out based on metadata values (55 samples across 28 patients).
- This left us with 118 samples across 59 patients. This is the data I have used in the below sequencing metrics analysis.
- The differential expression analysis includes the s341 samples (both primary and recurrent). This is not present in the sequencing metrics file.


```{r importlibraries, include=FALSE}

library(dplyr)

```

```{r importdata, include=FALSE}
seqdat <- read.delim("/Users/catherinehogg/Documents/Semester3/Project/Scripts/isoformproject/local/localdata/SequencingMetrics.txt", header = TRUE)
allbam <- read.delim("/Users/catherinehogg/Documents/Semester3/Project/Other/all_bam_files.txt", header = FALSE)
metadata <- read.csv("/Users/catherinehogg/Documents/Semester3/Project/Scripts/isoformproject/local/localdata/Metadata.csv", header = TRUE)
patients.remove <- read.delim("/Users/catherinehogg/Documents/Semester3/Project/Scripts/isoformproject/local/localdata/patients_remove.txt", header = FALSE)
```

```{r addpatientid, include=FALSE}

seqdat$patientid <- sub("_.*","",seqdat$Sample)
table(seqdat$patientid)
length(unique(seqdat$patientid))

dim(seqdat[is.na(seqdat$Sample.Source),])
```



```{r removedata, include=FALSE}

patients.remove <- as.vector(t(patients.remove))
idx <- !sub("_.*","",seqdat$Sample) %in% patients.remove
seqdat <- seqdat[idx,]

```

```{r datawrangling, include=FALSE}

seqdat$patientid <- sub("_.*","",seqdat$Sample)
seqdat <- left_join(seqdat, metadata, by = c("patientid" = "Patient.ID"))
seqdat$tumourtype <- sub(".*_","",seqdat$Sample)
seqdat$tumourtype <- sub("Primary","P",seqdat$tumourtype)
seqdat$tumourtype <- sub("Recurrent","R",seqdat$tumourtype)
seqdat <- seqdat[!is.na(seqdat$Sample.Source),]

seqdat$patientid <- as.factor(seqdat$patientid)
seqdat$tumourtype <- as.factor(seqdat$tumourtype)
seqdat$Sample.Source <- as.factor(seqdat$Sample.Source)

seqdat <- seqdat[order(seqdat$Sample.Source),]
```

```{r plotreads, echo=FALSE, fig.cap = "Read depth and length by sample source"}
plot(seqdat$Reads*1e-6, seqdat$ReadLength, col = seqdat$Sample.Source, pch = 16,
     xlab = "Read Depth (in millions)", ylab = "Read Length")
legend(105,200, legend = levels(seqdat$Sample.Source), col = c(1,2,3,4,5), pch = 16, cex = 0.8)
```

```{r plotmappedreads, echo=FALSE, fig.width = 10, fig.cap = "Number of mapped and unmapped reads. Total number of samples = 118 (59 patients). Samples grouped by source in the following order: Kim, Nam, Rabadan, Stead, Verhaak."}
barplot(as.matrix(t(cbind(seqdat$Reads*1e-6,(seqdat$Reads-seqdat$TotalMapped)*1e-6))),
        ylab = "Number of Reads (millions)", xlab = "Sample", names = seqdat$Sample, las = 3, cex.names = 0.3,
        cex.axis = 0.5, density = c(80,0), legend = c("Mapped Reads","Unmapped Reads"), args.legend = list(x = "topleft", cex = 1, bty = "n"))
```

```{r percentagereads, echo=FALSE, fig.cap = "Proportion unmapped reads"}

seqdat$unmapped_pct <- 1 - seqdat$TotalMapped/seqdat$Reads
seqdat$uniquelymapped_pct <- seqdat$UniquelyMapped/seqdat$Reads
seqdat$multimapped_pct <- seqdat$MultiMapped/seqdat$Reads

plot(seqdat$unmapped_pct, col = seqdat$Sample.Source, pch = 16, xlab = "Sample", ylab = "Proportion unmapped reads", ylim = c(0,0.45), xaxt = "null")
legend(0,0.4, legend = levels(seqdat$Sample.Source), col = c(1,2,3,4,5), pch = 16, cex = 0.8)

```

```{r percentagereads2, echo=FALSE, fig.cap = "Proportion multi-mapped reads"}
plot(seqdat$multimapped_pct, col = seqdat$Sample.Source, pch = 16, xlab = "Sample", ylab = "Proportion multi-mapped reads", ylim = c(0,0.4), xaxt = "null")
legend(100,0.4, legend = levels(seqdat$Sample.Source), col = c(1,2,3,4,5), pch = 16, cex = 0.8)

```

```{r sampleslessthan30m}

below30 <- as.vector(unique(seqdat[seqdat$Reads < 30000000,c(17)]))
below30
write.table(below30,"/Users/catherinehogg/Documents/Semester3/Project/Scripts/isoformproject/local/localdata/below30.txt",sep = "\t",row.names = FALSE, col.names = FALSE)
```
